src/Destructurama.JsonNet/JsonNet/JsonNetDestructuringPolicy.cs

```diff
@@ -41,22 +41,29 @@ public bool TryDestructure(object value, ILogEventPropertyValueFactory propertyV
             return false;
         }
 
-        LogEventPropertyValue Destructure(JValue jv, ILogEventPropertyValueFactory propertyValueFactory)
+        private static LogEventPropertyValue Destructure(JValue jv, ILogEventPropertyValueFactory propertyValueFactory)
         {
             return propertyValueFactory.CreatePropertyValue(jv.Value, true);
         }
 
-        // ReSharper disable once ParameterTypeCanBeEnumerable.Local
-        LogEventPropertyValue Destructure(JArray ja, ILogEventPropertyValueFactory propertyValueFactory)
+        private static LogEventPropertyValue Destructure(JArray ja, ILogEventPropertyValueFactory propertyValueFactory)
         {
             var elems = ja.Select(t => propertyValueFactory.CreatePropertyValue(t, true));
             return new SequenceValue(elems);
         }
 
-        LogEventPropertyValue Destructure(JObject jo, ILogEventPropertyValueFactory propertyValueFactory)
+        private static LogEventPropertyValue Destructure(JObject jo, ILogEventPropertyValueFactory propertyValueFactory)
+        {
+            return jo.Properties().All(prop => LogEventProperty.IsValidName(prop.Name)) ?
+                 DestructureToStructureValue(jo, propertyValueFactory) :
+                 DestructureToDictionaryValue(jo, propertyValueFactory);
+        }
+
+        private static LogEventPropertyValue DestructureToStructureValue(JObject jo, ILogEventPropertyValueFactory propertyValueFactory)
         {
             string typeTag = null;
             var props = new List<LogEventProperty>(jo.Count);
+
             foreach (var prop in jo.Properties())
             {
                 if (prop.Name == "$type")
```

Comment:
If we add an:

```csharp
else if (!LogEventProperty.IsValidName(prop.Name)) {
    return DestructureToDictionaryValue(jo, propertyValueFactory);
}
```

We can potentially avoid the `All()` call in the `Destructure()` method above and save the allocation of an enumerator plus the cost of the traversal of the properties (`MoveNext()` etc.). It's probably not very expensive, but seems like a quick and easy win that still reflects the basic purpose of validating property names and falling back if an invalid one is detected. What do you think?

Reply:
Yes good idea, I did not think of this approach. Was tossing between what I eventually did, and this:
- Iterate through `jo.Properties()` and populate both `props` and `elements`
- Check `IsValidName` on every iteration. If found one, stop checking, and also stop populating `props`.
- After the loop, return either `DictionaryValue` from `elements`, or `StructureValue` from `props`.

Out of the 3 approaches, yours is probably the slowest if `JObject` does contain an invalid property (particularly at the end of the list), but it's the fastest if no invalid props. Presuming most `JObject`s do not contain invalid props, your approach should be optimal. Let me do some perf testing...

Reply:
:+1:

Reply:
Hi Andrew, thanks for the numbers!

I don't think your conclusion necessarily fits the spread of expected inputs - so far, every single successful log statement running through this functionality "in the wild" is on path 1, so the interesting numbers are:

> 1. No invalid properties.
> Mine: 534 ms
> Yours: 513 ms (4% faster)

The only logging statement that will benefit from choosing the optimal implementation for 2, 3, or 4, is the one that's currently rejected, wouldn't you agree? :-)

(4% is probably splitting hairs, but it's often hard to see the impact of allocations in microbenchmarks, which is why the code tends to avoid enumerators in many places today.)

Reply:
@nblumhardt Ok, all done